## COSC 1210
### print(), input(), value storage and chaos

Objectives
- use the chaos function to illustrate the importance of value storage

In [4]:

def main():
    a = 1/7
    b = a + a + a + a + a + a + a
    print(b)
main()

0.9999999999999998


##### Explain why the function above provides an answer that is different than the algebraic solution
This is because the computer does not round an answer. So the answer will only keep getting closer to 1, never actually reaching the answer.


In [5]:
def main():
    a = 1/7
    b = a * 7
    c = a + a + a + a + a + a + a
    print("b is", b)
    print("c is", c)
    print(b-c)
main()

b is 1.0
c is 0.9999999999999998
2.220446049250313e-16


##### Explain why b and c provide different answers
b and c are different because when adding fractions the computer will never actually reach the whole number. When the function is mulitplied the computer does rounding. It is built in.


##### How large is the difference between c and b?
The difference between c and b is very small. About .1 or smaller.



### rounding floats
Mathematically, what will be the value of b?
When calculated by python, what is the value of b?

In [ ]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 10
    x1 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
    print("For", iterations, "iterations, x1 is:", x1)
main()

##### If you calculate x1 for 10 iterations vs 100 iterations, what answers do you get?

for 10 iterations: 0.9735732406265619

for 100 iterations: 0.5126527786207501

##### What are these different?
The number changes each time though the calculation.

In [1]:
def main():
    print("This program illustrates a chaotic function")
    iterations = 10
    x1 = 0.15
    x2 = 0.15
    for i in range(iterations):
        x1 = 3.9 * x1 * (1 - x1)
        x2 = 3.9 * (x2 - x2 * x2)
    print("For", iterations, "iternations, x1 is now", x1, "and x2 is now", x2)
main()

This program illustrates a chaotic function
For 10 iternations, x1 is now 0.9735732406265619 and x2 is now 0.9735732406265659


##### If you calculate x1 and x2 for 10 iterations vs 100 iterations, what answers do you get?

x1 10 iterations: 0.9735732406265619

x2 10 iterations: 0.9735732406265659



x1 100 iterations:0.5126527786207501

x2 100 iterations: 0.16411537941360743

##### Why are these different?
The order of operations is what gives them different answers. Because the variable is used multiple times of x2, the rounding gets extremely off. So when there are less variables, it leaves less room for rounding to become off.

In [6]:
### Rewrite the chaos function so that the use can enter x
### Rewrite the chaos function so that the use can enter x
def main():
    print("This program illustrates a chaotic function")
    user_num= input()
    iteration = user_num